0006-bingings.md

```diff
@@ -0,0 +1,591 @@
+* **Promise API.** Recently we have discussed the changes to our promise mechanics. The schema does not need to change,
+but the specification now needs to be clarified;
+* `data_read` currently has mixed functionality -- it can be used both for reading data from trie and to read data from
+the context. In former it expects pointers to be passed as arguments, in later it expects enum to be passed. It achieves
+juxtaposition by casting pointer type in enum when needed;
+* **Economics API.** The functions that provide access to balance and such might need to be added or removed since we
+now consider splitting attached balance into two.
+
+# Specification
+
+## Scratch Buffer
+Scratch buffer allows the host function to return the data into a buffer located inside the host oppose to the buffer
+located on the client. A special operation can be used to copy the content of the buffer into the host. Memory pointers
+can then be used to point either to the memory on the guest or the memory on the host, see below. Benefits:
+* We can have functions that return values that are not necessarily used, e.g. inserting key-value into a trie can
+also return the preempted old value, which might not be necessarily used. Previously, if we returned something we
+would have to pass the blob from host into the guest, even if it is not used;
+* We can pass blobs of data between host functions without going through the guest, e.g. we can remove the value
+from the storage and insert it into under a different key;
+* It makes API cleaner, because we don't need to pass `buffer_len` and `buffer_ptr` as arguments to other functions;
+* It allows merging certain functions together, see `storage_iter_next`;
+* This is consistent with other APIs that were created for high performance, e.g. allegedly Ewasm have implemented
+SNARK-like computations in Wasm by exposing a bignum library through stack-like interface to the guest. The guest
+can manipulate then with the stack of 256-bit numbers that is located on the host.
+
+#### Host → host blob passing
+The scratch buffer can be used to pass the blobs between host functions. For any function that
+takes a pair of arguments `*_len: u64, *_ptr: u64` this pair is pointing to a region of memory either on the guest or
+the host:
+* If `*_len != u64::MAX` it points to the memory on the guest;
+* If `*_len == u64::MAX` it points to the memory under the register `*_ptr` on the host.
+
+For example:
+`storage_write(u64::MAX, 0, u64::MAX, 1, 2)` -- insert key-value into storage, where key is read from register 0,
+value is read from register 1, and result is saved to register 2.
+
+#### Specification
+```rust
+read_register(register_id: u64, ptr: u64)
+```
+Writes the entire content from the scratch buffer `register_id` into the memory of the guest starting with `ptr`.
+###### Panics
+* If the content extends outside the memory allocated to the guest. In Wasmer, it returns `MemoryAccessViolation` error message;
```

Comment:
Added a comment.

This is pretty much true for any API that deals with raw pointers. We cannot avoid raw pointers with Wasm because we need to work with blobs of data of arbitrary size. The advantage of registers is that we now have one function that can corrupt the guest memory. Oppose to having multiple specialized functions like `read_account_id(ptr: u64)`  that can do that.